node/libs/concurrency/src/sync/prunable_queue.rs

```diff
@@ -0,0 +1,148 @@
+//! Prunable queue mimics a producer-consumer channel, in enabling
+//! waiting on the receiver side for new items, while providing pruning functionality.
+//! The pruning takes place whenever a new item is sent, based on a specified predicate.
+//!
+//! The queue facilitates the asynchronous return of processing result for each added item.
+//!
+//! The separation of `Sender` and `Receiver` is employed primarily because `Receiver` requires
+//! a mutable reference to the signaling channel, unlike `Sender`, hence making it undesirable to
+//! be used in conjunction.
+
+use crate::{
+    ctx::{self, Ctx},
+    oneshot,
+    sync::{self, watch},
+};
+use std::{collections::VecDeque, sync::Arc};
+use tokio::sync::Mutex;
+
+pub fn new<T, U>(
+    pruning_predicate: Box<dyn Sync + Send + Fn(&T, &T) -> bool>,
+) -> (Sender<T, U>, Receiver<T, U>) {
+    let queue: Mutex<VecDeque<(T, oneshot::Sender<U>)>> = Mutex::new(VecDeque::new());
+    // Internal signaling, to enable waiting on the receiver side for new items.
+    let (send, recv) = watch::channel(false);
+
+    let shared = Arc::new(Shared {
+        queue,
+        has_items_send: send,
+    });
+
+    let sender = Sender {
+        shared: shared.clone(),
+        pruning_predicate: pruning_predicate,
+    };
+
+    let receiver = Receiver {
+        shared: shared.clone(),
+        has_items_recv: recv,
+    };
+
+    return (sender, receiver);
+}
+
+pub struct Shared<T, U> {
+    queue: Mutex<VecDeque<(T, oneshot::Sender<U>)>>,
+    has_items_send: watch::Sender<bool>,
+}
+
+pub struct Sender<T, U> {
+    shared: Arc<Shared<T, U>>,
+    pruning_predicate: Box<dyn Sync + Send + Fn(&T, &T) -> bool>,
+}
+
+impl<T, U> Sender<T, U> {
+    pub async fn enqueue(&self, item: T) -> oneshot::Receiver<U> {
+        // Create oneshot channel for returning result asynchronously.
+        let (res_send, res_recv) = oneshot::channel();
+
+        let mut queue = self.shared.queue.lock().await;
+        queue.retain(|existing_item| !(self.pruning_predicate)(&existing_item.0, &item));
+        queue.push_back((item, res_send));
```

Comment:
what if the new message is older than a message already in the queue?

Reply:
Currently, won't be pruned. I wasn't sure about this.

Reply:
it should be pruned, as it opens an attack vector of getting spammed with messages in decreasing view order.

Reply:
Changed here: https://github.com/matter-labs/era-consensus/pull/56/commits/677b4677e24ebc4d9624214808492b32ff3b2b1a.
It disallow messages from the same type from the same validator altogether.

Reply:
that's also not ok, because the validator should be allowed to override their previous message if our node is stuck processing. You may argue that it should be allowed to override it unconditionally, but that complicates reasoning about the bft algorithm.

Reply:
Sorry if it wasn't clear, the pruning/disallowing always refer to the pending messages. New messages are always accepted, hence overriding is supported. 

Not sure I understood your last sentence though.

Reply:
ok,that's acceptable semantics then.